Docsp 22414 (#357)
mongodb/docs-node: source/fundamentals/encrypt-fields.txt

==============
Encrypt Fields
==============

.. default-domain:: mongodb

Overview
--------

You can encrypt fields in a document using a feature called
**client-side encryption**.

Client-side encryption enables your client applications to encrypt data
*before* sending it to MongoDB, and to query documents for values in
encrypted fields.

Because the driver encrypts the data before sending it to MongoDB, only
your configured client applications can decrypt the data. People or
applications with access to your encryption keys are the only ones that
can see the decrypted, plaintext data. Should you have unauthorized access
to your database, and attacker would only see the encrypted version of the
fields you configured for encryption.

Client-side encryption is useful for storing sensitive data, such as:

- credit card numbers
- addresses
- health information
- financial information
- any other sensitive or personally identifiable information (PII)

> MongoDB offers two ways to encrypt data:

Author: Good point. I don't expect this to change, so I quantified the list. I think the second suggestion enumerates it exactly and would need revision if another fle feature is introduced in the future.

==============
Encrypt Fields
==============

.. default-domain:: mongodb

> 

Author: It should, good callout.

Overview
--------

You can encrypt fields in a document using a feature called
**client-side encryption**.

Client-side encryption enables your client applications to encrypt data
*before* sending it to MongoDB, and to query documents for values in
encrypted fields.

Because the driver encrypts the data before sending it to MongoDB, only

> your configured client applications can decrypt the data. People or
> applications with access to your encryption keys are the only ones that
> can see the decrypted, plaintext data. Should you have unauthorized access

Reviewer: ```suggestion
your configured client applications can decrypt the data. 
Only applications using the driver with access to your encryption keys 
can access the decrypted, plaintext data. Should you have unauthorized access
```

to your database, an attacker could only see the encrypted version of the
fields you configured for encryption.

Client-side can prevent exposure of the following sensitive types of data:

- Credit card numbers
- Addresses
- Health information
- Financial information
- Any other sensitive or personally identifiable information (PII)

MongoDB offers the following ways to encrypt data:

{+qe+}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

{+qe+} is the next-generation client-side encryption feature,
introduced in MongoDB 6.0. {+qe+} supports searching encrypted fields
for equality and encrypts each value uniquely.

> The MongoDB manual contains detailed information for {+qe+}:

Reviewer: ```suggestion
The MongoDB manual contains detailed information on the following {+qe+} topics:
```

- Begin a working {+qe+} example with the :ref:`qe-quick-start`
- Learn how to use {+qe+} in the :ref:`qe-fundamentals`
- Learn how to integrate your {+qe+} implementation with a KMS in the :ref:`qe-tutorials`
- Learn {+qe+} concepts in the :ref:`qe-reference`

{+csfle-long+}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

{+csfle-long+} supports searching encrypted values.

> The core difference between {+csfle-short+} and {+qe+} is that {+csfle-short+}
> requires that encrypted fields you want to search must be determinstically

Reviewer: ```suggestion
{+csfle-short+} differs from {+qe+} in that it
requires encrypted fields you want to search be determinstically
```

encrypted. When you encrypt a value with deterministic encryption, the
same input value produces the same output value.

An unauthorized party who gains access to your data may be able to infer
the plaintext values of fields that contain common values. For example,
an attacker could infer the ``occupation`` field for all documents with
a value of "software engineer" if they know that value for a specific
document.

> The MongoDB manual contains detailed information for CSFLE:

Reviewer: ```suggestion
The MongoDB manual contains detailed information on the following {+csfle-short+} topics:
```

==============
Encrypt Fields
==============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. default-domain:: mongodb

Overview
--------

You can encrypt fields in a document using a feature called

> **client-side encryption**.

Reviewer 2: since we call it In-Use encryption in the other docs we should also use that here.  Maybe call it a set of features called In-Use Encryption so that it references more than one?

Client-side encryption enables your client applications to encrypt data

> *before* sending it to MongoDB, and to query documents for values in

Reviewer 2: only equality lets you query for values, so perhaps just "query documents with encrypted fields"?

encrypted fields.

Because the driver encrypts the data before sending it to MongoDB, only
your configured client applications can decrypt the data. Only applications
using the driver with access to your encryption keys can access the decrypted,
plaintext data. Should you have unauthorized access
to your database, an attacker could only see the encrypted version of the
fields you configured for encryption.

> Client-side encryption can prevent exposure of the following sensitive types of data:

Reviewer 2: I'd say "help prevent" or something similar, since they could do many things at the app side that would expose it

- Credit card numbers
- Addresses
- Health information
- Financial information
- Any other sensitive or personally identifiable information (PII)

> MongoDB offers the following ways to encrypt data:

Reviewer 2: ..to encrypt fields (since at-rest also encrypts data)

Author: good point!

{+qe+}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

{+qe+} is the next-generation client-side encryption feature,
introduced in MongoDB 6.0. {+qe+} supports searching encrypted fields
for equality and encrypts each value uniquely.

The MongoDB manual contains detailed information on the following {+qe+} topics:

- Begin a working {+qe+} example with the :ref:`qe-quick-start`
- Learn how to use {+qe+} in the :ref:`qe-fundamentals`
- Learn how to integrate your {+qe+} implementation with a KMS in the :ref:`qe-tutorials`
- Learn {+qe+} concepts in the :ref:`qe-reference`

{+csfle-long+}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> {+csfle-long+} was introduced in MongoDB v4.2 and supports searching encrypted fields.

Reviewer 2: since you mention for equality above we should say it here too

{+csfle-short+} differs from {+qe+} in that it requires that encrypted fields
you want to search must be determinstically encrypted. When you deterministically
encrypt a value, the same input value produces the same output value.

> An unauthorized party who gains access to your data may be able to infer

Reviewer 2: This is too strong a statement about leakage and inference because it is only fields with low cardinality that is a problem so I'd take some version of the wording from the deterministic description in the docs "While deterministic encryption provides greater support for read operations, encrypted data with low cardinality is susceptible to frequency analysis recovery."

Reviewer 2: so i don't want an example in there either, just the statement about about low cardinality
